### DISCUSSION TITLE
Development setup

### DISCUSSION THREAD

**Discussion by jarrah42** [silx-kit/h5web#1334] [Q&A] [answered]
Hi, can you describe how to set up h5web for local development? I'm trying to write my own provider, then test the local version using a local install of my galaxy visualization plugin, but I cannot get my plugin to build. The visualization plugin uses yarn, and builds fine when I use the npm @h5web packages.

I am able to build the h5web packages from source using `pnpm packages` and I get app/lib/shared directories under packages.  To use the local version of h5web I use the commands:

```
yarn add file:/path/to/packages/app
yarn add file:/path/to/packages/lib
yarn add file:/path/to/packages/shared
```

Then I run `yarn install; yarn build` but it gives me the error:

```
yarn run v1.22.19
$ parcel build src/script.js -d static --no-source-maps --detailed-report
🚨  /Users/gw6/galaxy/config/plugins/visualizations/h5web/node_modules/@h5web/app/src/App.tsx:31:12: This experimental syntax requires enabling one of the following parser plugin(s): "jsx", "flow", "typescript". (31:12)
```

I don't understand why it is trying to build using the typescript file and not using the javascript version. The only thing I've noticed is that the `package.json` in app uses `"main":"src/index.ts"` while the npm version uses `"main":"dist/index.js"` but I don't know why this would be or how to change this. I have been trying to figure this out for weeks to no avail. Any help would be appreciated.

**Comment by loichuder** [comment]
Hello,

Thanks for your interest and sorry to hear this stumped you for a while. @axelboc is the main person to ask for the setup but I'll try to help you with the best of my ability.

> The only thing I've noticed is that the package.json in app uses "main":"src/index.ts" while the npm version uses "main":"dist/index.js" but I don't know why this would be or how to change this

This is done by pnpm [`publishConfig` field](https://pnpm.io/package_json#publishconfig) in the `package.json`. The fields in `publishConfig` replace the actual fields in the `package.json` when publishing the package, hence the change you noticed.

A quick-and-dirty fix may be to manually replicate this behaviour by editing the `package.json` of the `h5web` packages of your local h5web setup (that is replacing the fields in `package.json` by their values of `publishConfig`).

If that doesn't work, we will be sure to get back to you after the holiday season. :santa:

**Comment by axelboc** [comment, marked as answer]
> This is done by pnpm [publishConfig field](https://pnpm.io/package_json#publishconfig) in the package.json. The fields in publishConfig replace the actual fields in the package.json when publishing the package, hence the change you noticed.

> A quick-and-dirty fix may be to manually replicate this behaviour by editing the package.json of the h5web packages of your local h5web setup (that is replacing the fields in package.json by their values of publishConfig).

@loichuder is spot on. If you'd like to automate this instead of having to edit the `package.json` files manually, you could try using pnpm's [`pack` command](https://pnpm.io/cli/pack):

1. `pnpm packages`
2. `cd packages/lib`
3. `pnpm pack --pack-destination <dir>`
4. This creates a tarball with the exact same content as the package published to NPM, so you can then extract it where you need, install it in your project with `yarn add`, do the same with the other packages, and run your build.

**Reply by jarrah42** [reply]
Hi all,

Thanks for the information. I wasn't able to get `pack` to work for me as it just seems to tar up the directory and it doesn't look anything like what I get when I install using NPM. In any case, here's what I ended up doing in case anyone else wants to use a similar method.

1. Install and start `verdaccio`.
2. Add `registry=http://localhost:4873/` to `~/.npmrc` so that `verdaccio` will be used as a cache.
3. After making a change to the h5web source code, run `pnpm packages` from the top level directory.
4. Change to `packages/lib` and modify `package.json` to remove the `"type": "module"` line.
5. Run `pnpm unpublish @h5web/lib@<version>` if already published. I made `<version>` something that wasn't already available via NPM.
6. Run `pnpm publish --no-git-checks`.
7. Restore the  `"type": "module"` line in `package.json`.
8. Repeat 4-7 in `packages/app` if necessary.

I made a simple script to do 3-8 that I could run each time I made a change.

In the project that uses h5web, run `yarn remove @h5web/lib` and `yarn add @h5web/lib` (and the same for `@h5web/app` if necessary) each time the package is rebuilt.

**Reply by axelboc** [reply]
Right, we do remove `"type": "module"` before publishing: https://github.com/silx-kit/h5web/blob/main/.github/workflows/release.yml#L66-L72 -- I didn't make the connection sorry. Perhaps this can be moved to a `"prepublish"` script on each package.

**Reply by axelboc** [reply]
I do wonder why `pnpm pack` didn't work, though. Did you run it from the correct directory - i.e. the root directory of each package like `packages/lib`? Also, there's an extra `package` folder in the tarball, but the content of that folder should be exactly the same as the content of the package on NPM. It could also be the version of pnpm -- are you on the latest?

**Reply by axelboc** [reply]
With the latest commits on `main`, you should now be able to remove step 4 from your script. This step will now be performed automatically when running `pnpm publish` (and `pnpm pack`). Let us know how it goes.

**Reply by jarrah42** [reply]
Yes now I can build/publish with the same `package.json` which makes it a lot easier. Thanks!

**Reply by axelboc** [reply]
FYI, I've added a _Local testing_ section to the CONTRIBUTING guide.: https://github.com/silx-kit/h5web/blob/main/CONTRIBUTING.md#local-testing. It mentions using `pnpm pack` (but not publishing to a local registry).
